core/src/Kokkos_MathematicalFunctions.hpp

```diff
@@ -0,0 +1,232 @@
+  KOKKOS_IMPL_UNARY_FUNCTION_FLOATING_POINT(nan, double, char const*)
+
+#else  // long double overloads are available
+
+#define KOKKOS_IMPL_BINARY_FUNCTION_ARITHMETIC(FUNC)                         \
+  template <typename Arithmetic1, typename Arithmetic2,                      \
+            typename =                                                       \
+                std::enable_if_t<std::is_arithmetic<Arithmetic1>::value &&   \
+                                 std::is_arithmetic<Arithmetic2>::value>,    \
+            typename Promoted = std::conditional_t<                          \
+                std::is_same<Arithmetic1, long double>::value ||             \
+                    std::is_same<Arithmetic2, long double>::value,           \
+                long double, double>>                                        \
+  KOKKOS_INLINE_FUNCTION Promoted FUNC(Arithmetic1 x, Arithmetic2 y) {       \
+    return Kokkos::Experimental::FUNC(                                       \
+        static_cast<std::conditional_t<std::is_integral<Arithmetic1>::value, \
+                                       double, Arithmetic1>>(x),             \
+        static_cast<std::conditional_t<std::is_integral<Arithmetic2>::value, \
+                                       double, Arithmetic2>>(y));            \
+  }
+
+#define KOKKOS_IMPL_MATH_UNARY_FUNCTION(FUNC)                               \
+  KOKKOS_IMPL_UNARY_FUNCTION_FLOATING_POINT(FUNC, float, float)             \
+  KOKKOS_IMPL_UNARY_FUNCTION_FLOATING_POINT(FUNC, double, double)           \
+  KOKKOS_IMPL_UNARY_FUNCTION_FLOATING_POINT(FUNC, long double, long double) \
+  KOKKOS_IMPL_UNARY_FUNCTION_INTEGRAL(FUNC, double)
+
+#define KOKKOS_IMPL_MATH_UNARY_PREDICATE(FUNC)                       \
+  KOKKOS_IMPL_UNARY_FUNCTION_FLOATING_POINT(FUNC, bool, float)       \
+  KOKKOS_IMPL_UNARY_FUNCTION_FLOATING_POINT(FUNC, bool, double)      \
+  KOKKOS_IMPL_UNARY_FUNCTION_FLOATING_POINT(FUNC, bool, long double) \
+  KOKKOS_IMPL_UNARY_FUNCTION_INTEGRAL(FUNC, bool)
+
+#define KOKKOS_IMPL_MATH_BINARY_FUNCTION(FUNC)                  \
+  KOKKOS_IMPL_BINARY_FUNCTION_FLOATING_POINT(FUNC, float)       \
+  KOKKOS_IMPL_BINARY_FUNCTION_FLOATING_POINT(FUNC, double)      \
+  KOKKOS_IMPL_BINARY_FUNCTION_FLOATING_POINT(FUNC, long double) \
+  KOKKOS_IMPL_BINARY_FUNCTION_ARITHMETIC(FUNC)
+
+#define KOKKOS_IMPL_MATH_NAN()                                        \
+  KOKKOS_IMPL_UNARY_FUNCTION_FLOATING_POINT(nanf, float, char const*) \
+  KOKKOS_IMPL_UNARY_FUNCTION_FLOATING_POINT(nan, double, char const*) \
+  KOKKOS_IMPL_UNARY_FUNCTION_FLOATING_POINT(nanl, long double, char const*)
+
+#endif
+
+// Basic operations
+KOKKOS_IMPL_MATH_UNARY_FUNCTION(fabs)
+KOKKOS_IMPL_MATH_BINARY_FUNCTION(fmod)
+KOKKOS_IMPL_MATH_BINARY_FUNCTION(remainder)
+KOKKOS_IMPL_MATH_BINARY_FUNCTION(fmin)
+KOKKOS_IMPL_MATH_BINARY_FUNCTION(fmax)
+KOKKOS_IMPL_MATH_BINARY_FUNCTION(fdim)
+#ifndef KOKKOS_ENABLE_SYCL
+KOKKOS_IMPL_MATH_NAN()
```

Comment:
`cl::sycl::nan` takes a `nancode` integer instead of a sequence of characters as defined everywhere else

Reply:
Ugh.  I agree with leaving it out of the SYCL builds for now.  I've filed an issue with Khronos on it (to add a `double nan(const char*)` overload to SYCL), although it is very unlikely to make it into SYCL 2020.

Reply:
`nan(const char*)` calls `__builtin_nan(const char*)` on both clang and gcc (not just under SYCL but in general).

Assuming it works on the device (haven't yet tried that), what would be correct way to add this in w/o waiting for dpc++ or SYCL to adopt it?

Reply:
Not use the macro for `nan` and do it "by hand"

Reply:
I would prefer to tackle this in follow up PR